### DISCUSSION TITLE
Error test suite

### DISCUSSION THREAD

**Discussion by Virgiel** [cornucopia-rs/cornucopia#59] [Ideas]
It might be interesting to have a test suite that checks that we still catch errors after refactoring. I've started experimenting with this in my [fork](https://github.com/Virgiel/cornucopia/tree/test_suite).

My idea is to write some Toml files describing inputs and the expected error:
```toml
[[test]]
name = "PgParamInExtendedQuery"
query = """
--! new_author
INSERT INTO Author (id, name) VALUES (:id, $1);
"""
error = """
Error while reading queries [path: "queries/module_1.sql"]:  --> 2:45
  | 
  | INSERT INTO Author (id, name) VALUES (:id, $1);
  |                                             ^---
  | 
  = Indexed bind parameters (`$index`) are not allowed when using the extended syntax.
  = Consider using a named bind parameter like `:identifier`, or use the PostgreSQL-compatible syntax..
"""

[[test]]
name = "ColumnNameAlreadyTaken"
migration = """
CREATE TABLE item (
  name TEXT,
  price INTEGER
);
"""
query = """
--! items
SELECT name, price AS name FROM item;
"""
error = """
Error while preparing query "items" [file: "queries/module_1.sql", line: 1]:
Two or more columns have the same name: `name`. Consider disambiguing the column names with `AS` clauses.
"""
```
Then we have a test runner that generates a temporary file tree with the specified queries and migrations, runs cornucopia CLI on it and checks the returned error. By testing at the boundaries, we can change the underlying code without changing the tests and when we change the error message, we should have an option for the test runner to automatically update the error message in the test descriptor.

By the way, your error messages are really good and user-friendly!

Do you find this useful? Would you prefer something simpler or no test at all for this?

**Comment by LouisGariepy** [comment]
That would be fantastic. I'd accept such a PR for sure :smile:, even if it was in a simpler/restricted form.

**Reply by Virgiel** [reply]
Done #61
